`ColorChooser` : TMI Sliders

ericmehl opened the pull request:
This adds three new color sliders to the `ColorChooser` : temperature, magenta and intensity. The conversion between TMI and RGB is based on the formulas from https://www.comp-fu.com/2012/05/tmi-color-temperature-correction/ and modified to account for the different between grading an existing color and generating one.

This is the first in a few PRs adding features to `ColorChooser`. Our intention is to not expose the changes until a final PR to incorporate all the changes, because the layout in particular is likely to evolve as additional features are added. The last commit is meant as a testing commit to test this PR. We will drop it before actually merging.

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -44,24 +44,55 @@
+)
+__rgb2tmiMatrix = __tmi2rgbMatrix.inverse()
+
+def _tmi2rgb( c ) :
```
`tmiToRGB` and `rgbToTMI` perhaps?

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5962/commits/025b13f3e7761a8e6e3147b22ffb1f8e728548bf

johnhaddon commented on python/GafferUITest/ColorChooserTest.py:
```diff
@@ -0,0 +1,92 @@
+
+import IECoreImage
+
+import Gaffer
+import GafferUI
```
Unused imports

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5962/commits/57811dc840731f87c8b6d56d59b326374640e33b

johnhaddon commented on python/GafferUITest/ColorChooserTest.py:
```diff
@@ -0,0 +1,92 @@
+		# left to right, varies the intensity from 0.0 to 0.9.
+		tileWidth = 20
+		tileCount = 10
+		tmiTarget = IECoreImage.ImageReader( ( pathlib.Path( __file__ ).parent / "images" / "tmi.exr" ).as_posix() ).read()
```
I'd like to avoid any further dependencies on ImageReader - we're getting close to being able to ditch it completely. Can we use OpenImageIO instead?

johnhaddon replied:
Alternately, we could perhaps test without an image - my vague recollection was that that was the plan anyway.

ericmehl replied:
Unfortunately OpenImageIO tries to return image pixels as a numpy array and fails if it can't do so. I don't see any way to ask it for a regular array or access pixels one at a time.

I could do a brute-force script to serialize the colors from the EXR directly into Python to include in the test script. Or maybe write out a file just made up of the floats and read it as standard binary data?

johnhaddon replied:
Is [getchannel](https://openimageio.readthedocs.io/en/latest/imagebuf.html#_CPPv4NK4OIIO8ImageBuf10getchannelEiiii8WrapMode) available in Python?

ericmehl replied:
Ah yes, that does the job. I'm using `getpixel()` in https://github.com/GafferHQ/gaffer/pull/5962/commits/846d22fe6cda3bf21a98cc6a50bc299f952ab53d because `getchannel()` was complaining about mismatched arguments even when I used the example from their documentation.

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -158,23 +158,23 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
+		self.__channelLabels = {}
 		self.__componentValueChangedConnections = []
 
 		with self.__column :
```
Maybe better to deal with this as part of a future layout change, but perhaps not we have a grid layout, we could use that for everything rather than nest everything in a column?

ericmehl replied:
I think I have enough questions about what the layout will become that I'll defer this until later, just to keep the code revision down.

johnhaddon commented on python/GafferUI/ColorChooser.py:
```diff
@@ -159,14 +159,20 @@ def __init__( self, color=imath.Color3f( 1 ), **kw ) :
+				components = "rgb_a_hsv_tmi"
+				for row, component in enumerate( components ) :
+					if component == "_" :
+						self.__channelSpacers[components[row - 1]] = GafferUI.Spacer( imath.V2i( 0, 4 ), imath.V2i( 0, 4 ), parenting = { "index" : ( 0, row ) } )
```
I know we're going to drop this before merging, but here's an observation to feed into the layout design : it does feel potentially useful to provide some grouping in some way, but the spacing feels out of balance with the surrounding widgets when embedded in the NodeEditor, and with the swatches when shown in a floating editor. The slider groups are more closely related to each other than the widgets around them, yet they have bigger spacing.

ericmehl replied:
Good point on the spacing, I see what you mean. I'll keep that in mind when we explore the layout options.

johnhaddon reviewed (commented):
Thanks Eric - seems to work well. I don't have much experience of using TMI sliders so I was surprised at how easy it is to end up with negative RGB while modifying TMI, but I see that a certain other package doesn't protect against that, so maybe folks will be savvy to that already?

ericmehl commented:
Thanks John! I addressed the comments, and stuck my head in the sand for a bit longer on the column vs. grid-only layout question, if that's ok.

It is odd how some components go outside the 0-1 range fairly easily. I'm not sure what a better behavior would be, the other DCCs I've tested allow it, as you noted, so I guess artists just need to be aware of it.

Maybe we could consider some kind of warning indicator when a channel goes outside the 0-1 range?

Ready for a new look!

ericmehl commented:
> I think this subtle indicator is probably enough compared to anything more drastic

That looks good to me, nice to keep it consistent with how it's been shown already.

I dropped the temporary commit in anticipation of merging. ( I have it locally for later or if you want me to add it back in for more testing). And yes, merging this to 1.4 was my intention, thanks for catching that.

johnhaddon commented:
Looks like the previous commits have a dependency on the temp commit.

ericmehl commented:
It was attempting to set the slider value for a slider that didn't exist. Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/5962/commits/f8f094514075f2acb567f33e4b5cd718c54fc0d6. Should be good to go now.
